File: day16/day16.kt
import java.io.File

fun <E> List<List<E>>.transpose(): List<List<E>> {
    return (0 until first().size).map { col ->
        (0 until size).map { row -> this[row][col] }
    }
}

fun main() {
    val content = File("day16.txt").readText().trimEnd().split("\n\n")
    val rules = Regex("""([\w ]+): (\d+)-(\d+) or (\d+)-(\d+)""").findAll(content[0]).map {
        val (r1, r2) = it.groupValues.drop(2).map { it.toInt() }.chunked(2).map { (a, b) -> (a..b) }
        it.groupValues[1] to r1.toSet() + r2.toSet()
    }.toMap()
    val parse = { s: String -> s.split('\n').drop(1).map { it.split(',').map { it.toInt() } } }
    val yourTicket = parse(content[1])[0]
    val nearbyTickets = parse(content[2])

    val values = rules.values.flatten().toSet()
    val validTickets = nearbyTickets.filter { it.all { it in values } }
    println(nearbyTickets.flatten().filter { it !in values }.sum())

    val options = validTickets.transpose().mapIndexed { k, v ->
        k to rules.entries.mapNotNull { (name, valid) ->
            if (v.all { it in valid }) name else null
        }
    }.sortedBy { it.second.size }
    val names = options.fold(mutableMapOf<String,Int>()) {
        m, (k, v) -> m[v.filter { it !in m }[0]] = k; m
    }

    val departure = rules.keys.filter { it.startsWith("departure") }.map { yourTicket[names[it]!!] }
    println(departure.fold(1L) { a, b -> a * b })
}

File: 2020/day16/day16.py
import re

# parse input
content = open('day16.txt').read().rstrip().split('\n\n')
rules = {m[0]: set(v for i in range(1, 5, 2) for v in range(int(m[i]), int(m[i+1]) + 1))
         for m in re.findall(r'^(.+): (\d+)-(\d+) or (\d+)-(\d+)', content[0], re.M)}
your = list(map(int, content[1].split('\n')[1].split(',')))
nearby = [list(map(int, line.split(','))) for line in content[2].split('\n')[1:]]

# detect invalid
values = set(v for s in rules.values() for v in s)
valid = [t for t in nearby if all(v in values for v in t)]
print(sum(v for t in nearby for v in t if v not in values))

# resolve mapping
options = [[k for k, v in rules.items() if all(x in v for x in a)] for a in zip(*valid)]
ordered = sorted(enumerate(options), key=lambda x: len(x[1]))
names = {}
for k, v in ordered:
    names[[s for s in v if s not in names][0]] = k

# calculate result
result = 1
for k, v in names.items():
    if k.startswith('departure'):
        result *= your[v]
print(result)
